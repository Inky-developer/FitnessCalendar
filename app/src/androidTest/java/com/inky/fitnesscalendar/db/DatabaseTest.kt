package com.inky.fitnesscalendar.db

import android.content.Context
import androidx.room.Room
import androidx.room.RoomDatabase
import androidx.sqlite.db.SupportSQLiteDatabase
import androidx.test.core.app.ApplicationProvider
import androidx.test.ext.junit.runners.AndroidJUnit4
import kotlinx.coroutines.runBlocking
import org.junit.Assert.assertEquals
import org.junit.Assert.assertNotNull
import org.junit.Test
import org.junit.runner.RunWith

@RunWith(AndroidJUnit4::class)
class DatabaseTest {
    /**
     * Tests that the default data get loaded correctly
     */
    @Test
    fun test_load_default_data() {
        val context = ApplicationProvider.getApplicationContext<Context>()
        val db = Room
            .inMemoryDatabaseBuilder(context, AppDatabase::class.java)
            .addCallback(object : RoomDatabase.Callback() {
                override fun onCreate(db: SupportSQLiteDatabase) {
                    loadDefaultData(db, context)
                }
            })
            .build()

        val activityTypeDao = db.activityTypeDao()
        val activityTypes = runBlocking { activityTypeDao.loadTypes() }
        assertEquals(activityTypes.size, DefaultActivityType.entries.size)
        for (activityType in DefaultActivityType.entries) {
            assertNotNull(
                "Activity type got created",
                activityTypes.find { it.name == context.getString(activityType.titleId) })
        }
    }
}