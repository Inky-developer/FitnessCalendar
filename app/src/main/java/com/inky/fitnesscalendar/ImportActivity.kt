package com.inky.fitnesscalendar

import android.content.Intent
import android.net.Uri
import android.os.Bundle
import android.os.Parcelable
import android.util.Log
import androidx.activity.ComponentActivity
import androidx.activity.compose.setContent
import androidx.activity.enableEdgeToEdge
import androidx.activity.viewModels
import com.inky.fitnesscalendar.ui.ImportView
import com.inky.fitnesscalendar.ui.theme.FitnessCalendarTheme
import com.inky.fitnesscalendar.ui.util.ProvideDatabaseValues
import com.inky.fitnesscalendar.view_model.ImportViewModel
import dagger.hilt.android.AndroidEntryPoint
import java.io.FileNotFoundException

private const val TAG = "ImportActivity"

@AndroidEntryPoint
class ImportActivity : ComponentActivity() {
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        enableEdgeToEdge()

        val viewModel: ImportViewModel by viewModels()
        viewModel.closeActivity = {
            finish()
        }

        if (savedInstanceState == null) {
            val uris = when (intent.action) {
                Intent.ACTION_SEND -> handleSendIntent()
                Intent.ACTION_SEND_MULTIPLE -> handleSendMultipleIntent()
                Intent.ACTION_VIEW -> handleViewIntent()
                else -> throw RuntimeException("Unexpected action ${intent.action}")
            }

            Log.i(TAG, "Received: $uris")

            val files = try {
                uris.mapNotNull { contentResolver.openFileDescriptor(it, "r") }
            } catch (e: FileNotFoundException) {
                Log.e(TAG, "File not found: $e")
                finish()
                return
            }

            if (files.isEmpty()) {
                Log.e(TAG, "No valid files in the intent")
                finish()
                return
            }

            viewModel.loadFiles(files)
        }

        setContent {
            FitnessCalendarTheme {
                ProvideDatabaseValues(repository = viewModel.repository) {
                    ImportView(viewModel)
                }
            }
        }
    }

    private fun handleSendIntent(): List<Uri> {
        @Suppress("DEPRECATION") val uri = intent.getParcelableExtra(Intent.EXTRA_STREAM) as? Uri
            ?: return emptyList()

        return listOf(uri)
    }

    private fun handleSendMultipleIntent(): List<Uri> {
        @Suppress("DEPRECATION") val uris =
            intent
                .getParcelableArrayListExtra<Parcelable>(Intent.EXTRA_STREAM)
                ?.mapNotNull { it as? Uri }
                ?: return emptyList()

        return uris
    }

    private fun handleViewIntent(): List<Uri> {
        return listOfNotNull(intent.data)
    }
}

