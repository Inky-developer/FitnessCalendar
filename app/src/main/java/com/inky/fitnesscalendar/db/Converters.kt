package com.inky.fitnesscalendar.db

import android.net.Uri
import androidx.room.TypeConverter
import com.inky.fitnesscalendar.data.EpochDay
import com.inky.fitnesscalendar.data.Intensity
import com.inky.fitnesscalendar.data.gpx.GpxTrackPoint
import com.inky.fitnesscalendar.data.measure.Distance
import kotlinx.serialization.encodeToString
import kotlinx.serialization.json.Json
import java.util.Date

class Converters {
    @TypeConverter
    fun dateToTimestamp(date: Date?): Long? {
        return date?.time
    }

    @TypeConverter
    fun dateFromTimestamp(value: Long?): Date? {
        return value?.let { Date(it) }
    }

    @TypeConverter
    fun uriToString(uri: Uri?): String? {
        return uri?.toString()
    }

    @TypeConverter
    fun uriFromString(value: String?): Uri? {
        return value?.let { Uri.parse(it) }
    }

    @TypeConverter
    fun epochDayToLong(day: EpochDay): Long {
        return day.day
    }

    @TypeConverter
    fun longToEpochDay(value: Long): EpochDay {
        return EpochDay(value)
    }

    @TypeConverter
    fun distanceToLong(distance: Distance): Long {
        return distance.meters
    }

    @TypeConverter
    fun longToDistance(value: Long): Distance {
        return Distance(meters = value)
    }

    @TypeConverter
    fun intensityToLong(intensity: Intensity): Byte {
        return intensity.value
    }

    @TypeConverter
    fun byteToIntensity(value: Byte): Intensity {
        return Intensity(value)
    }

    @TypeConverter
    fun trackPointsToByteArray(points: List<GpxTrackPoint>): ByteArray {
        return Json.encodeToString(points).toByteArray()
    }

    @TypeConverter
    fun byteArrayToTrackPoints(array: ByteArray): List<GpxTrackPoint> {
        return Json.decodeFromString(array.decodeToString())
    }
}