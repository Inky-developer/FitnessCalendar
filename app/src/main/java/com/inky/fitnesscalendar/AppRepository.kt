package com.inky.fitnesscalendar

import android.content.Context
import android.util.Log
import androidx.compose.runtime.Immutable
import androidx.room.withTransaction
import com.inky.fitnesscalendar.data.EpochDay
import com.inky.fitnesscalendar.data.Vehicle
import com.inky.fitnesscalendar.data.activity_filter.ActivityFilter
import com.inky.fitnesscalendar.data.activity_filter.ActivityFilterChip
import com.inky.fitnesscalendar.data.activity_filter.ActivityFilterChip.Companion.toActivityFilterChip
import com.inky.fitnesscalendar.db.AppDatabase
import com.inky.fitnesscalendar.db.dao.ActivityDao
import com.inky.fitnesscalendar.db.dao.ActivityTypeDao
import com.inky.fitnesscalendar.db.dao.DayDao
import com.inky.fitnesscalendar.db.dao.FilterHistoryDao
import com.inky.fitnesscalendar.db.dao.PlaceDao
import com.inky.fitnesscalendar.db.dao.RecordingDao
import com.inky.fitnesscalendar.db.entities.Activity
import com.inky.fitnesscalendar.db.entities.ActivityType
import com.inky.fitnesscalendar.db.entities.Day
import com.inky.fitnesscalendar.db.entities.Place
import com.inky.fitnesscalendar.db.entities.Recording
import com.inky.fitnesscalendar.db.entities.RichActivity
import com.inky.fitnesscalendar.db.entities.RichRecording
import com.inky.fitnesscalendar.di.ActivityTypeOrder
import com.inky.fitnesscalendar.localization.LocalizationRepository
import com.inky.fitnesscalendar.util.hideRecordingNotification
import com.inky.fitnesscalendar.util.showRecordingNotification
import dagger.hilt.android.qualifiers.ApplicationContext
import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.flow.first
import kotlinx.coroutines.flow.map
import javax.inject.Inject
import javax.inject.Singleton

private const val TAG = "AppRepository"

@Immutable
@Singleton
class AppRepository @Inject constructor(
    @ApplicationContext val context: Context,
    private val database: AppDatabase,
    private val activityDao: ActivityDao,
    private val recordingDao: RecordingDao,
    private val activityTypeDao: ActivityTypeDao,
    private val filterHistoryDao: FilterHistoryDao,
    private val dayDao: DayDao,
    private val placeDao: PlaceDao,
    val localizationRepository: LocalizationRepository
) {
    suspend fun loadAllActivities() = activityDao.loadActivities()

    fun getActivities(filter: ActivityFilter): Flow<List<RichActivity>> {
        // Use simpler query if no filters are needed
        if (filter.isEmpty()) {
            return activityDao.getActivities()
        }

        val searchVehicles = filter.text?.let { filterText ->
            Vehicle.entries.filter { vehicle ->
                context.getString(vehicle.nameId).contains(filterText, ignoreCase = true)
            }
        } ?: emptyList()

        val categories = filter.categories.map { it.toString() }

        val rangeDate = filter.range?.getDateRange()

        val hasDescription = filter.attributes.description.toBooleanOrNull()
        val hasFeel = filter.attributes.feel.toBooleanOrNull()
        val hasImage = filter.attributes.image.toBooleanOrNull()

        return activityDao.getFiltered(
            typeIds = filter.types.mapNotNull { it.uid },
            isTypesEmpty = filter.types.isEmpty(),
            categories = categories,
            isCategoriesEmpty = categories.isEmpty(),
            placeIds = filter.places.mapNotNull { it.uid },
            isPlacesEmpty = filter.places.isEmpty(),
            search = filter.text?.let { "%$it%" },
            searchVehicles = searchVehicles,
            start = rangeDate?.start,
            end = rangeDate?.end,
            hasDescription = hasDescription,
            hasFeel = hasFeel,
            hasImage = hasImage
        )
    }

    fun getMostRecentActivity() = activityDao.getMostRecentActivity()

    suspend fun saveActivity(activity: RichActivity) {
        activityDao.save(activity.clean().activity)
    }

    suspend fun deleteActivity(activity: Activity) {
        activityDao.delete(activity)
    }

    suspend fun loadMostRecentActivities(n: Int) = activityDao.loadMostRecentActivities(n)

    fun getActivity(id: Int) = activityDao.get(id)

    suspend fun getActivityImages() = activityDao.getImages()

    suspend fun startRecording(richRecording: RichRecording, context: Context) {
        val recordingId = recordingDao.insert(richRecording.recording).toInt()
        context.showRecordingNotification(
            recordingId,
            richRecording.type,
            richRecording.recording.startTime.time
        )
    }

    fun getRecordings() = recordingDao.getRecordings()

    suspend fun getRecording(uid: Int) = recordingDao.getById(uid)

    suspend fun deleteRecording(recording: Recording) {
        Log.d(TAG, "Deleting $recording")
        recording.uid?.let { context.hideRecordingNotification(it) }
        recordingDao.delete(recording)
    }

    suspend fun endRecording(recording: Recording) {
        Log.d(TAG, "Ending recording $recording")
        recording.uid?.let { context.hideRecordingNotification(it) }
        val type = getActivityType(recording.typeId)
        if (type == null) {
            Log.e(TAG, "Could not retrieve type for activity")
            return
        }
        val activity = recording.toActivity(type)
        activityDao.stopRecording(recording, activity)
    }

    suspend fun saveActivityType(activityType: ActivityType) = activityTypeDao.save(activityType)

    suspend fun deleteActivityType(activityType: ActivityType) =
        activityTypeDao.delete(activityType)

    suspend fun loadActivityTypes() = activityTypeDao.loadTypes()

    fun getActivityTypes() = activityTypeDao.getTypes()

    private suspend fun getActivityType(id: Int): ActivityType? = activityTypeDao.get(id)

    fun getActivityTypeRows() =
        getActivityTypesByCategory().map { ActivityTypeOrder.getRowsOrDefault(it) }

    private fun getActivityTypesByCategory() = activityTypeDao.getActivityTypesByCategory()

    suspend fun upsertFilterHistoryChips(chips: List<ActivityFilterChip>) =
        database.withTransaction {
            val historyItems = getFilterHistoryItems().first()
                .associate { it.toActivityFilterChip()!! to it.item.uid!! }
            for (item in chips) {
                val existingItemId = historyItems[item]
                filterHistoryDao.upsert(
                    item.toFilterHistoryItem().copy(uid = existingItemId)
                )
            }
            filterHistoryDao.onlyKeepNewest(8)
        }

    fun getFilterHistoryItems() = filterHistoryDao.getItems()

    fun getDays() = dayDao.getDays()

    fun getDay(day: EpochDay): Flow<Day> = dayDao.get(day).map { it ?: Day(day = day) }

    suspend fun saveDay(day: Day) {
        dayDao.upsert(day)
    }

    suspend fun savePlace(place: Place) = placeDao.upsert(place)

    suspend fun deletePlace(place: Place) = placeDao.delete(place)

    fun getPlaces() = placeDao.getAll()
}