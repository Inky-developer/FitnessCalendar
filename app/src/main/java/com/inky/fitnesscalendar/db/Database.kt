package com.inky.fitnesscalendar.db

import android.content.Context
import androidx.room.AutoMigration
import androidx.room.Database
import androidx.room.Room
import androidx.room.RoomDatabase
import androidx.room.TypeConverters
import androidx.sqlite.db.SupportSQLiteDatabase
import com.inky.fitnesscalendar.db.dao.ActivityDao
import com.inky.fitnesscalendar.db.dao.ActivityTypeDao
import com.inky.fitnesscalendar.db.dao.DayDao
import com.inky.fitnesscalendar.db.dao.FilterHistoryDao
import com.inky.fitnesscalendar.db.dao.PlaceDao
import com.inky.fitnesscalendar.db.dao.RecordingDao
import com.inky.fitnesscalendar.db.entities.Activity
import com.inky.fitnesscalendar.db.entities.ActivityType
import com.inky.fitnesscalendar.db.entities.Day
import com.inky.fitnesscalendar.db.entities.FilterHistoryItem
import com.inky.fitnesscalendar.db.entities.Place
import com.inky.fitnesscalendar.db.entities.Recording
import com.inky.fitnesscalendar.util.DATABASE_NAME

@Database(
    version = 17,
    entities = [Activity::class, Recording::class, ActivityType::class, FilterHistoryItem::class, Day::class, Place::class],
    autoMigrations = [
        AutoMigration(from = 2, to = 3),
        AutoMigration(from = 3, to = 4),
        AutoMigration(from = 4, to = 5, spec = Migration4To5Spec::class),
        AutoMigration(from = 10, to = 11),
        AutoMigration(from = 11, to = 12),
        AutoMigration(from = 13, to = 14),
        AutoMigration(from = 14, to = 15),
        AutoMigration(from = 15, to = 16),
        AutoMigration(from = 16, to = 17),
    ]
)
@TypeConverters(Converters::class)
abstract class AppDatabase : RoomDatabase() {
    abstract fun activityDao(): ActivityDao

    abstract fun recordingDao(): RecordingDao

    abstract fun activityTypeDao(): ActivityTypeDao

    abstract fun filterHistoryDao(): FilterHistoryDao

    abstract fun dayDao(): DayDao

    abstract fun placeDao(): PlaceDao

    companion object {
        @Volatile
        private var instance: AppDatabase? = null

        fun getInstance(context: Context): AppDatabase {
            return instance ?: synchronized(this) {
                instance ?: buildDatabase(context).also { instance = it }
            }
        }

        private fun buildDatabase(context: Context) =
            Room
                .databaseBuilder(context, AppDatabase::class.java, DATABASE_NAME)
                .addMigrations(
                    MIGRATION_1_2,
                    MIGRATION_5_6,
                    MIGRATION_6_7,
                    MIGRATION_7_8,
                    MIGRATION_8_9,
                    MIGRATION_9_10,
                    MIGRATION_12_13
                )
                .addCallback(object : Callback() {
                    override fun onCreate(db: SupportSQLiteDatabase) {
                        loadDefaultData(db, context)
                    }
                })
                .build()
    }
}