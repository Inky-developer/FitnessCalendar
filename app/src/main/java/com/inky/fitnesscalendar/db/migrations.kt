package com.inky.fitnesscalendar.db

import androidx.room.RenameColumn
import androidx.room.migration.AutoMigrationSpec
import androidx.room.migration.Migration
import androidx.sqlite.db.SupportSQLiteDatabase

val MIGRATION_1_2 = object : Migration(1, 2) {
    override fun migrate(db: SupportSQLiteDatabase) {
        db.execSQL("UPDATE Activity SET type = \"UniversityCommute\" WHERE type = \"WorkCommute\"")
    }
}

@RenameColumn(tableName = "Activity", fromColumnName = "type", toColumnName = "type_id")
class Migration4To5Spec: AutoMigrationSpec