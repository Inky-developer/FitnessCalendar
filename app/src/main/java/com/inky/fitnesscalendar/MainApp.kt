package com.inky.fitnesscalendar

import android.app.Application
import android.util.Log
import androidx.core.net.toUri
import com.inky.fitnesscalendar.di.ActivityTypeOrder
import com.inky.fitnesscalendar.di.DecisionTrees
import com.inky.fitnesscalendar.repository.AppRepository
import com.inky.fitnesscalendar.util.cleanImageStorage
import com.inky.fitnesscalendar.util.getOrCreateImagesDir
import com.inky.fitnesscalendar.util.getOrCreateSharedMediaCache
import dagger.hilt.android.HiltAndroidApp
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.MainScope
import kotlinx.coroutines.launch
import javax.inject.Inject

@HiltAndroidApp
class MainApp : Application() {
    @Inject
    lateinit var appRepository: AppRepository

    override fun onCreate() {
        super.onCreate()

        MainScope().launch(Dispatchers.IO) {
            initializeData()
            cleanupStorage()
        }
    }

    private suspend fun cleanupStorage() {
        val imageDir = getOrCreateImagesDir().toPath()
        val usedImages = appRepository.getUsedImages().map { imageDir.resolve(it).toFile().toUri() }
        cleanImageStorage(usedImages.toSet())

        getOrCreateSharedMediaCache().listFiles()?.forEach {
            it.delete()
        }
    }

    private suspend fun initializeData() {
        Log.i("MainApp", "Initializing app data")

        val activities = appRepository.loadMostRecentActivities(200)

        ActivityTypeOrder.init(activities)
        DecisionTrees.init(activities)
        Log.i("MainApp", "App data successfully initialized")
    }
}