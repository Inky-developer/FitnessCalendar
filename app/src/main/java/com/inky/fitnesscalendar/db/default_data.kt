package com.inky.fitnesscalendar.db

import android.content.Context
import android.os.Build
import androidx.annotation.RequiresApi
import androidx.annotation.StringRes
import androidx.sqlite.db.SupportSQLiteDatabase
import com.inky.fitnesscalendar.R
import com.inky.fitnesscalendar.data.ActivityCategory
import com.inky.fitnesscalendar.data.ContentColor
import com.inky.fitnesscalendar.data.measure.HeartFrequency
import com.inky.fitnesscalendar.data.measure.Temperature
import com.inky.fitnesscalendar.data.measure.meters
import com.inky.fitnesscalendar.db.entities.Activity
import com.inky.fitnesscalendar.util.toDate
import kotlinx.coroutines.runBlocking
import java.time.LocalDate
import java.time.Period
import java.time.ZoneId
import kotlin.random.Random

fun loadDefaultData(db: SupportSQLiteDatabase, context: Context) {
    db.beginTransaction()
    try {
        for (type in DefaultActivityType.entries) {
            db.execSQL(
                "INSERT INTO ActivityType(activity_category, name, emoji, color, has_vehicle, has_duration, has_distance, has_intensity) VALUES (?, ?, ?, ?, ?, ?, ?, ?)",
                arrayOf<Any>(
                    type.activityCategory,
                    context.getString(type.titleId),
                    type.emoji,
                    type.color,
                    type.hasVehicle,
                    type.hasDuration,
                    type.hasDistance,
                    type.hasIntensity
                )
            )
        }

        db.setTransactionSuccessful()
    } finally {
        db.endTransaction()
    }
}

enum class DefaultActivityType(
    val activityCategory: ActivityCategory,
    @StringRes val titleId: Int,
    val emoji: String,
    val color: ContentColor,
    val hasVehicle: Boolean = false,
    val hasDuration: Boolean = true,
    val hasDistance: Boolean = false,
    val hasIntensity: Boolean = false
) {
    Cycling(
        ActivityCategory.Sports,
        titleId = R.string.activity_cycling,
        color = ContentColor.Color2,
        emoji = "üö¥",
        hasDistance = true
    ),
    Running(
        ActivityCategory.Sports,
        titleId = R.string.activity_running,
        color = ContentColor.Color3,
        emoji = "üèÉ",
        hasDistance = true
    ),
    WorkCommute(
        ActivityCategory.Travel,
        titleId = R.string.activity_commute_home_to_work,
        color = ContentColor.Color1,
        emoji = "üè¢",
        hasVehicle = true
    ),
    HomeCommute(
        ActivityCategory.Travel,
        titleId = R.string.activity_commute_work_to_home,
        color = ContentColor.Color2,
        emoji = "üè°",
        hasVehicle = true
    ),
    Travel(
        ActivityCategory.Travel,
        titleId = R.string.travel,
        color = ContentColor.Color3,
        emoji = "üó∫Ô∏è",
        hasVehicle = true
    ),
    Work(
        ActivityCategory.Work,
        titleId = R.string.work,
        color = ContentColor.Color1,
        emoji = "üíº"
    ),
    Gaming(
        ActivityCategory.Entertainment,
        titleId = R.string.gaming,
        color = ContentColor.Color1,
        emoji = "üéÆ"
    ),
    Film(
        ActivityCategory.Entertainment,
        titleId = R.string.film,
        color = ContentColor.Color2,
        emoji = "üé¨"
    ),
    Note(
        ActivityCategory.Other,
        titleId = R.string.activity_note,
        color = ContentColor.Color1,
        emoji = "üìì",
        hasDuration = false
    ),
    HealthNote(
        ActivityCategory.Other,
        titleId = R.string.health_note,
        color = ContentColor.Color2,
        emoji = "ü©π",
        hasDuration = false
    ),
    Other(
        ActivityCategory.Other,
        titleId = R.string.other,
        color = ContentColor.Color3,
        emoji = "üè∑Ô∏è"
    );
}

/**
 * Produces some default activities for testing purposes
 */
@RequiresApi(Build.VERSION_CODES.UPSIDE_DOWN_CAKE)
fun generateSampleActivities(db: AppDatabase, seed: Int = 42) = runBlocking {
    val startDay = LocalDate.of(2025, 1, 1)
    val endDay = LocalDate.of(2025, 6, 1)

    val random = Random(seed)

    val activityTypes = db.activityTypeDao().loadTypes()

    val sports = activityTypes.filter { it.name == "Cycling" || it.name == "Running" }

    for (day in startDay.datesUntil(endDay, Period.ofDays(2))) {
        val startTime = day.atTime(random.nextInt(19), random.nextInt(60))
        val endTime = startTime.plusMinutes(random.nextLong(60 * 4))
        val activityType = sports.random(random)
        val activity = Activity(
            typeId = activityType.uid!!,
            startTime = startTime.toDate(ZoneId.of("UTC")),
            endTime = endTime.toDate(ZoneId.of("UTC")),
            description = "Some ${activityType.name}",
            favorite = random.nextDouble() < 0.05,
            distance = random.nextDouble(8000.0, 80000.0).meters(),
            temperature = Temperature(celsius = random.nextDouble(-10.0, 40.0)),
            averageHeartRate = HeartFrequency(bpm = random.nextDouble(120.0, 170.0)),
            maximalHeartRate = HeartFrequency(bpm = random.nextDouble(150.0, 200.0)),
        )
        db.activityDao().save(activity)
    }
}