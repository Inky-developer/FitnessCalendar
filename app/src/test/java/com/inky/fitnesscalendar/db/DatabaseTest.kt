package com.inky.fitnesscalendar.db

import android.os.Build
import androidx.room.Room
import androidx.room.RoomDatabase
import androidx.sqlite.db.SupportSQLiteDatabase
import androidx.test.core.app.ApplicationProvider
import com.inky.fitnesscalendar.MainApp
import kotlinx.coroutines.runBlocking
import org.approvaltests.Approvals
import org.junit.Test
import org.junit.runner.RunWith
import org.robolectric.RobolectricTestRunner
import org.robolectric.annotation.Config

@Config(sdk = [Build.VERSION_CODES.UPSIDE_DOWN_CAKE])
@RunWith(RobolectricTestRunner::class)
class DatabaseTest {
    /**
     * Tests that the default data get loaded correctly
     */
    @Test
    fun test_load_default_data() {
        val context = ApplicationProvider.getApplicationContext<MainApp>()
        val db = Room
            .inMemoryDatabaseBuilder(context, AppDatabase::class.java)
            .addCallback(object : RoomDatabase.Callback() {
                override fun onCreate(db: SupportSQLiteDatabase) {
                    loadDefaultData(db, context)
                }
            })
            .build()

        val activityTypeDao = db.activityTypeDao()
        val activityTypes = runBlocking { activityTypeDao.loadTypes() }
        Approvals.verifyAll("ActivityTypes", activityTypes)

        db.close()
    }
}