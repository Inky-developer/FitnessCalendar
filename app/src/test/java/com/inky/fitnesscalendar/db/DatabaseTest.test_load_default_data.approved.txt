ActivityTypes[0] = Cycling
ActivityTypes[1] = Running
ActivityTypes[2] = Home → Work
ActivityTypes[3] = Work → Home
ActivityTypes[4] = Travel
ActivityTypes[5] = Work
ActivityTypes[6] = Gaming
ActivityTypes[7] = Film
ActivityTypes[8] = Note
ActivityTypes[9] = Health
ActivityTypes[10] = Other
